# Data Understanding

## Objective
Understand dataset structure, data types, missing values, and overall quality.

## Approach
- Use generalized functions
- Avoid hardcoding
- Use loops for automation

In [2]:
#importing necessary libraries
import pandas as pd
import numpy as np

In [ ]:
#loading the dataset 
df =pd.read_csv("../data/raw/automobile.csv", header=None, na_values="?") #the dataset has no column names and has missing values
#displaying first 5 rows
df.head()

,0,1,2,3,4,5,6,7,8,9,...,16,17,18,19,20,21,22,23,24,25
0,3,NaN,alfa-romero,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111.0,5000.0,21,27,13495.0
1,3,NaN,alfa-romero,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111.0,5000.0,21,27,16500.0
2,1,NaN,alfa-romero,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154.0,5000.0,19,26,16500.0
3,2,164.0,audi,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102.0,5500.0,24,30,13950.0
4,2,164.0,audi,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115.0,5500.0,18,22,17450.0


In [4]:
#List of column names (from dataset documentation)
columns = [
    "symboling", "normalized_losses", "make", "fuel_type",
    "aspiration", "num_of_doors", "body_style", "drive_wheels",
    "engine_location", "wheel_base", "length", "width", "height",
    "curb_weight", "engine_type", "num_of_cylinders", "engine_size",
    "fuel_system", "bore", "stroke", "compression_ratio",
    "horsepower", "peak_rpm", "city_mpg", "highway_mpg", "price"
]
#Assigning column names to dataframe
df.columns = columns

#check updated dataset
df.head()

,symboling,normalized_losses,make,fuel_type,aspiration,num_of_doors,body_style,drive_wheels,engine_location,wheel_base,...,engine_size,fuel_system,bore,stroke,compression_ratio,horsepower,peak_rpm,city_mpg,highway_mpg,price
0,3,NaN,alfa-romero,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111.0,5000.0,21,27,13495.0
1,3,NaN,alfa-romero,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111.0,5000.0,21,27,16500.0
2,1,NaN,alfa-romero,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154.0,5000.0,19,26,16500.0
3,2,164.0,audi,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102.0,5500.0,24,30,13950.0
4,2,164.0,audi,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115.0,5500.0,18,22,17450.0


In [ ]:
# Function to display basic information about dataset
def dataset_overview(df):
    
    # Print shape of dataset (rows, columns)
    print("Shape:", df.shape)
    
    # Print column names
    print("\nColumns:\n", df.columns.tolist())
    
    # Print data types of each column
    print("\nData Types:\n", df.dtypes)

In [12]:
#running the above function
dataset_overview(df)

Shape: (205, 26)

Columns:
 ['symboling', 'normalized_losses', 'make', 'fuel_type', 'aspiration', 'num_of_doors', 'body_style', 'drive_wheels', 'engine_location', 'wheel_base', 'length', 'width', 'height', 'curb_weight', 'engine_type', 'num_of_cylinders', 'engine_size', 'fuel_system', 'bore', 'stroke', 'compression_ratio', 'horsepower', 'peak_rpm', 'city_mpg', 'highway_mpg', 'price']

Data Types:
 symboling              int64
normalized_losses    float64
make                  object
fuel_type             object
aspiration            object
num_of_doors          object
body_style            object
drive_wheels          object
engine_location       object
wheel_base           float64
length               float64
width                float64
height               float64
curb_weight            int64
engine_type           object
num_of_cylinders      object
engine_size            int64
fuel_system           object
bore                 float64
stroke               float64
compression_ratio  

In [13]:
# Function to calculate missing values in each column
def missing_values(df):
    
    # Count missing values column-wise
    missing = df.isnull().sum()
    
    # Return only columns where missing values exist (> 0)
    return missing[missing > 0]

In [14]:
missing_values(df)

normalized_losses    41
num_of_doors          2
bore                  4
stroke                4
horsepower            2
peak_rpm              2
price                 4
dtype: int64

In [15]:
# Function to count unique values in each column
def unique_values(df):
    
    result = {}  # Dictionary to store results
    
    # Loop through each column
    for col in df.columns:
        
        # Count unique values in column
        result[col] = df[col].nunique()
    
    return result

In [16]:
unique_values(df)

{'symboling': 6,
 'normalized_losses': 51,
 'make': 22,
 'fuel_type': 2,
 'aspiration': 2,
 'num_of_doors': 2,
 'body_style': 5,
 'drive_wheels': 3,
 'engine_location': 2,
 'wheel_base': 53,
 'length': 75,
 'width': 44,
 'height': 49,
 'curb_weight': 171,
 'engine_type': 7,
 'num_of_cylinders': 7,
 'engine_size': 44,
 'fuel_system': 8,
 'bore': 38,
 'stroke': 36,
 'compression_ratio': 32,
 'horsepower': 59,
 'peak_rpm': 23,
 'city_mpg': 29,
 'highway_mpg': 30,
 'price': 186}

In [17]:
# Function to separate numerical and categorical columns
def get_column_types(df):
    
    # Select numerical columns
    numeric_cols = df.select_dtypes(include=np.number).columns.tolist()
    
    # Select categorical columns
    categorical_cols = df.select_dtypes(exclude=np.number).columns.tolist()
    
    return numeric_cols, categorical_cols


# Call function
num_cols, cat_cols = get_column_types(df)

print("Numerical Columns:\n", num_cols)
print("\nCategorical Columns:\n", cat_cols)

Numerical Columns:
 ['symboling', 'normalized_losses', 'wheel_base', 'length', 'width', 'height', 'curb_weight', 'engine_size', 'bore', 'stroke', 'compression_ratio', 'horsepower', 'peak_rpm', 'city_mpg', 'highway_mpg', 'price']

Categorical Columns:
 ['make', 'fuel_type', 'aspiration', 'num_of_doors', 'body_style', 'drive_wheels', 'engine_location', 'engine_type', 'num_of_cylinders', 'fuel_system']


In [19]:
#Display statistical summary of numerical columns 
df.describe()

,symboling,normalized_losses,wheel_base,length,width,height,curb_weight,engine_size,bore,stroke,compression_ratio,horsepower,peak_rpm,city_mpg,highway_mpg,price
count,205.000000,164.000000,205.000000,205.000000,205.000000,205.000000,205.000000,205.000000,201.000000,201.000000,205.000000,203.000000,203.000000,205.000000,205.000000,201.000000
mean,0.834146,122.000000,98.756585,174.049268,65.907805,53.724878,2555.565854,126.907317,3.329751,3.255423,10.142537,104.256158,5125.369458,25.219512,30.751220,13207.129353
std,1.245307,35.442168,6.021776,12.337289,2.145204,2.443522,520.680204,41.642693,0.273539,0.316717,3.972040,39.714369,479.334560,6.542142,6.886443,7947.066342
min,-2.000000,65.000000,86.600000,141.100000,60.300000,47.800000,1488.000000,61.000000,2.540000,2.070000,7.000000,48.000000,4150.000000,13.000000,16.000000,5118.000000
25%,0.000000,94.000000,94.500000,166.300000,64.100000,52.000000,2145.000000,97.000000,3.150000,3.110000,8.600000,70.000000,4800.000000,19.000000,25.000000,7775.000000
50%,1.000000,115.000000,97.000000,173.200000,65.500000,54.100000,2414.000000,120.000000,3.310000,3.290000,9.000000,95.000000,5200.000000,24.000000,30.000000,10295.000000
75%,2.000000,150.000000,102.400000,183.100000,66.900000,55.500000,2935.000000,141.000000,3.590000,3.410000,9.400000,116.000000,5500.000000,30.000000,34.000000,16500.000000
max,3.000000,256.000000,120.900000,208.100000,72.300000,59.800000,4066.000000,326.000000,3.940000,4.170000,23.000000,288.000000,6600.000000,49.000000,54.000000,45400.000000


In [22]:
#count duplicate rows
print("Duplicate values: ",df.duplicated().sum())

Duplicate values:  0


##  Observations

- The dataset contains 205 rows and 26 columns representing automobile specifications and pricing.
- Both numerical and categorical features are present, requiring different preprocessing techniques.
- Missing values are observed in multiple columns such as:
 - normalized_losses (41 missing values)
 - num_of_doors, bore, stroke, horsepower, peak_rpm, and price
- Some columns like 'make', 'engine_type', and 'fuel_system' have high cardinality (many unique values).
- Numerical features show a wide range of values (e.g., engine_size, horsepower), indicating possible outliers.
- No duplicate rows are present in the dataset.
- Some columns that should be numerical may currently be treated as categorical due to missing values.

## Key Data Issues

- Missing values represented earlier as '?' have been converted to NaN.
- Data type inconsistencies exist and need to be fixed during preprocessing.
- Certain features may require scaling and encoding.

## Conclusion

The dataset has been successfully explored and key structural and quality issues have been identified.  
The data is now ready for further preprocessing and exploratory analysis.

